src/Network/XMPP/Types.hs

```diff
@@ -261,9 +326,9 @@ data StanzaType
 
 data Stanza :: StanzaType -> * where
     MkMessage ::
-        { mFrom    :: Maybe (JID '[])
-        , mTo      :: JID '[]
-        , mId      :: String -- ^ Message 'from', 'to', 'id' attributes                              
+        { mFrom    :: Maybe (JID 'NodeResource)
+        , mTo      :: SomeJID
```

Comment:
If we add one more phantom type to `Stanza`, something that would mark the 'destination'... we could use a type family over `StanzaType` and that new destination type, to calculate exact type of the JID necessary.
I'm thinking this might let us have Stanzas without wrapped JID type.

Actually, there's a `MessageType` and its carried as value inside MkMessage constructor. If we pull this one into a type, I think type family over those two would be able to select correct JID.

Eg.
```
type family MessageToJID (st :: StanzaType) (mt :: MessageType) :: JIDQualification where
    'Message 'Chat = 'Resource    -- Not sure how to use JIDQualification properly
    'Message 'GroupChat = 'NodeResource
     ....
```
and then in the Stanza
```
...
, mTo :: JID (MessageToJID 'Message mt)
...
```

Reply:
I am not sure, that `from` and `to` are always the same for each message type, and not only for outgoing, but for incoming :( 
I saw some examples in spec, where message has one jid for `to`, but another `from` (I mean different number of components) :(

Reply:
And, as I see, `from` and `to` are different from case to case at least in responses, so I am really unsure in rules of composition of kids for particular messages :( that's why I leaved it as `SomeJID`: just to avoid  some problems with parsing in some cases, while we have problems bigger :)

Reply:
As an example, here (request - has fully qualified jid, but response is not in `from` and `to`):

<img width="743" alt="image" src="https://user-images.githubusercontent.com/58581226/71249281-5ee9f180-234f-11ea-86a1-b3e86e7262f5.png">

Reply:
Yeah, this is actually precisely why I wanted to improve JIDs :-)
Before you added GADT, user of the library had no idea how to construct a JID for specific type of message, and would be forced to go look into XMPP specification to learn.
Now with GADTs, we refine JIDs on the components (ps - tiny bit of documentation on JIDQualification, or a link to a spec anchor would be great!) such as Resource, NodeResource and so on. I was thinking about encoding the *purpose* of the JID into type, so from your examples:
- in Example1, `iqFrom` JID would be of some type that represents format "Nickname@Server/MaybeResource", for example, `JID '[ name = 'Nick, server = 'Server, 'Resource = 'Nothing ]`
- Ex1, iqTo would be `JID '[ name = 'Nothing, server = 'Server, resource = 'Nothing ]`

The actual data layout -imo- wouldn't even matter much and original data layout would adopt all of this; you'd specify what kind of JID you need in your type, and then only examine fields that are guaranteed to be filled in. The rest can as well be undefined.

Althought, to be fair - GADTs really make it much nicer with requiring you to fill in only the fields you need, and yes the amount of permutations might end up being a lot - but I think its worth the effort. This is the exact type of knowledge that'd be worth embedding into the type logic, since its coming directly from the spec. Specific type of message needs specific type of JID, and we know there are three components to a JID: nickname, server and resource. What we need, aside a way to express their presence or absence, is a way to specify that in some instances, nickname will contain the *server* type, and resource will contain the *nickname* - this is exactly what happens when you join the room, look:
![2019-12-20-071400_583x223_scrot](https://user-images.githubusercontent.com/57259245/71251355-83ff4580-22f8-11ea-8a32-00018ab0b9b1.png)

Note how user `hag66` on the server `shakespeare.lit`, using her PDA, wants to join a room `coven` on the server `chat.shakespeare.lit` with a nickname `thirdwitch`.